### PR TITLE
fix: number of Prometheus replicas can be 0

### PR DESCRIPTION
**Pull request by tremes**
This updates my original PR cbb95f32509ecde346b70eb30bd558483f162e2f to allow 0 number of replicas for the Prometheus config.

### CODE REVIEW

**Review by simonpasquier — DISMISSED** [review]

**Comment by simonpasquier** [comment]
> then one still has to scale down the Prometheus operator (to scale Prometheus to custom number of replicas - including 0)

it shouldn't be needed. As soon as OBO set `spec.replicas`  to 0 for the Prometheus object, the Prometheus operator will scale the number of Prometheus pods to zero.

I'll defer the final approval to @sthaha and @jan--f :)

**Comment by sthaha** [comment]
@tremes I agree with Simon, that we shouldn't have to create artificial limits for the minimum allowed. IMHO, we should permit all values that PO permits this keeps the interfaces consistent. 
Could we please add a  test for the scenario that Simon mentioned that setting 0 does indeed scale down the pods. E.g. 

* set replica to 1
* validate pod count = 1
* set replica to 0
* validate pod count = 0

**Review by sthaha — CHANGES_REQUESTED** [review]
This looks good to me! 
Requesting change only for a test to be added that assert setting 0 does indeed scale down the pods.

File: test/e2e/monitoring_stack_controller_test.go [hunk 1]
```diff
@@ -517,6 +518,33 @@ func assertAlertmanagerReceivesAlerts(t *testing.T) {
...
+
+	key := types.NamespacedName{Name: ms.Name, Namespace: ms.Namespace}
+	err = f.K8sClient.Get(context.Background(), key, ms)
+	assert.NilError(t, err, "failed to get a monitoring stack")
+
+	numOfRep = 0
+	ms.Spec.PrometheusConfig.Replicas = &numOfRep
+	err = f.K8sClient.Update(context.Background(), ms)
+	assert.NilError(t, err, "failed to update a monitoring stack")
+	f.GetResourceWithRetry(t, ms.Name, ms.Namespace, &prom)
```
Comment (sthaha) [inline test/e2e/monitoring_stack_controller_test.go:544]: shouldn't we retry if the status takes time to update and then fail?

Comment (simonpasquier) [reply test/e2e/monitoring_stack_controller_test.go:544]: I assume that we want at least that generation == observedGeneration (for the MonitoringStack object) before retrieving the Prometheus object?

**Review by jan--f — APPROVED** [review]
This lgtm, will leave it to @sthaha to merge though.

File: test/e2e/monitoring_stack_controller_test.go [hunk 2]
```diff
@@ -517,6 +518,45 @@ func assertAlertmanagerReceivesAlerts(t *testing.T) {
...
+	err := f.K8sClient.Create(context.Background(), ms)
+	assert.NilError(t, err, "failed to create a monitoring stack")
+
+	prom := monv1.Prometheus{}
+	f.GetResourceWithRetry(t, ms.Name, ms.Namespace, &prom)
+
+	assert.Equal(t, prom.Status.Replicas, int32(1))
+
+	key := types.NamespacedName{Name: ms.Name, Namespace: ms.Namespace}
+	err = f.K8sClient.Get(context.Background(), key, ms)
```
Comment (sthaha) [inline test/e2e/monitoring_stack_controller_test.go:537]: nit: it is always safer to use a new blank object.

**Review by sthaha — APPROVED** [review]
lgtm, thank you for the patch @tremes 🙇

**Pull request merged**